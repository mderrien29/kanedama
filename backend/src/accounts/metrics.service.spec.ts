import { Test, TestingModule } from '@nestjs/testing';

import { MetricsService } from './metrics.service';
import { AccountDto } from 'src/common/dtos/account.dto';
import * as userTransactions from '../../test/resources/user-transactions.json';
import * as userAccounts from '../../test/resources/user-accounts.json';

jest.useFakeTimers('modern').setSystemTime(new Date('2021-12-01').getTime());

describe('MetricsService', () => {
  let metricsService: MetricsService;

  beforeEach(async () => {
    const app: TestingModule = await Test.createTestingModule({
      providers: [MetricsService],
    }).compile();

    metricsService = app.get<MetricsService>(MetricsService);
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  describe('getUserMetrics', () => {
    it('should return the expected result', () => {
      const userMetrics = metricsService.getUserMetrics(
        userAccounts as AccountDto[],
        userTransactions,
      );

      expect(userMetrics).toMatchSnapshot();
    });
  });

  describe('calculateAccountAmountOverTime', () => {
    it('should return the expected result', () => {
      const amountOverTime = metricsService.calculateAccountAmountOverTime(
        userTransactions,
        userAccounts as AccountDto[],
      );

      expect(amountOverTime).toMatchSnapshot();
    });
  });

  describe('calculateMinMaxBalance', () => {
    it('should return the expected result for test case 1', () => {
      const { min, max } = metricsService.calculateMinMaxBalance([
        0, 1, 2, 3, 3, 2, 1, 0,
      ]);

      expect(min).toMatchSnapshot();
      expect(max).toMatchSnapshot();
    });

    it('should return the expected result for test case 2', () => {
      const { min, max } = metricsService.calculateMinMaxBalance([0, -1, 100]);

      expect(min).toMatchSnapshot();
      expect(max).toMatchSnapshot();
    });
  });

  describe('hadRecentActivityLastYears', () => {
    it('should return true for last 10 years', () => {
      expect(
        metricsService.hadRecentActivityLastYears(userTransactions, 10),
      ).toBeTruthy();
    });

    it('should return false for last year', () => {
      expect(
        metricsService.hadRecentActivityLastYears(userTransactions, 1),
      ).toBeFalsy();
    });

    it('should return false if no transactions', () => {
      expect(metricsService.hadRecentActivityLastYears([], 10)).toBeFalsy();
    });
  });

  describe('calculateAverageIncomeLastMonths', () => {
    it('should return the average last 6 months', () => {
      const average = metricsService.calculateAverageIncomeLastMonths(
        userTransactions,
        6,
      );

      expect(average).toMatchSnapshot();
    });

    it('should return the average last 2 months', () => {
      const average = metricsService.calculateAverageIncomeLastMonths(
        userTransactions,
        2,
      );

      expect(average).toMatchSnapshot();
    });
  });
});
