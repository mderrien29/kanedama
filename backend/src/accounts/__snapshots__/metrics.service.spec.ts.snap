// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MetricsService calculateAccountAmountOverTime should return the expected result 1`] = `
Array [
  10000,
  10100,
  10095,
  9995,
  10015,
  10005,
  9005,
]
`;

exports[`MetricsService calculateAverageIncomeLastMonths should return the average last 2 months 1`] = `5`;

exports[`MetricsService calculateAverageIncomeLastMonths should return the average last 6 months 1`] = `38.333333333333336`;

exports[`MetricsService calculateMinMaxBalance should return the expected result for test case 1 1`] = `0`;

exports[`MetricsService calculateMinMaxBalance should return the expected result for test case 1 2`] = `3`;

exports[`MetricsService calculateMinMaxBalance should return the expected result for test case 2 1`] = `-1`;

exports[`MetricsService calculateMinMaxBalance should return the expected result for test case 2 2`] = `100`;

exports[`MetricsService getUserMetrics should return the expected result 1`] = `
Object {
  "3_years_activity": true,
  "6_month_average_income": 38,
  "max_balance": 10100,
  "min_balance": 9005,
}
`;
